cmd/soroban-cli/src/commands/init.rs

```diff
@@ -0,0 +1,329 @@
+    from: &Path,
+    to: &Path,
+    contracts: &[ExampleContract],
+) -> Result<(), Error> {
+    let project_contracts_path = to.join("contracts");
+    for contract in contracts {
+        let contract_as_string = contract.to_string();
+        let contract_path = Path::new(&contract_as_string);
+        let from_contract_path = from.join(contract_path);
+        let to_contract_path = project_contracts_path.join(contract_path);
+        std::fs::create_dir_all(&to_contract_path)?;
+
+        copy_contents(&from_contract_path, &to_contract_path)?;
+        edit_contract_cargo_file(&to_contract_path)?;
+    }
+
+    Ok(())
+}
+
+fn edit_contract_cargo_file(contract_path: &Path) -> Result<(), Error> {
+    let cargo_path = contract_path.join("Cargo.toml");
+    let cargo_toml_str = read_to_string(&cargo_path)?;
+    let mut doc = cargo_toml_str.parse::<Document>().unwrap();
+
+    let mut workspace_table = InlineTable::new();
+    workspace_table.insert("workspace", Value::Boolean(Formatted::new(true)));
+
+    doc["dependencies"]["soroban-sdk"] =
+        toml_edit::Item::Value(Value::InlineTable(workspace_table.clone()));
+    doc["dev_dependencies"]["soroban-sdk"] =
+        toml_edit::Item::Value(Value::InlineTable(workspace_table));
+
+    doc.remove("profile");
+
+    std::fs::write(&cargo_path, doc.to_string())?;
+
+    Ok(())
+}
+
+fn include_example_contracts(contracts: &[ExampleContract]) -> bool {
+    !(contracts.len() == 1 && contracts[0] == ExampleContract::None)
+}
+
+fn clone_repo(from_url: &str, to_path: &Path) -> Result<(), Error> {
+    let mut fetch = gix::clone::PrepareFetch::new(
+        from_url,
+        to_path,
+        gix::create::Kind::WithWorktree,
+        gix::create::Options {
+            destination_must_be_empty: false,
+            fs_capabilities: None,
+        },
+        gix::open::Options::isolated(),
+    )
+    .map_err(Box::new)?
+    .with_shallow(gix::remote::fetch::Shallow::DepthAtRemote(
+        NonZeroU32::new(1).unwrap(),
+    ));
+
+    let (mut prepare, _outcome) = fetch
+        .fetch_then_checkout(gix::progress::Discard, &AtomicBool::new(false))
+        .map_err(Box::new)?;
+
+    let (_repo, _outcome) =
+        prepare.main_worktree(gix::progress::Discard, &AtomicBool::new(false))?;
+
+    Ok(())
+}
+
+fn copy_contents(from: &Path, to: &Path) -> Result<(), Error> {
+    let contents_to_exclude_from_copy = [".git", ".github", "Makefile", "Cargo.lock", ".vscode"];
+    for entry in fs::read_dir(from)? {
+        let entry = entry?;
+        let path = entry.path();
+        let entry_name = entry.file_name().to_string_lossy().to_string();
+        let new_path = to.join(&entry_name);
+
+        if contents_to_exclude_from_copy.contains(&entry_name.as_str()) {
+            continue;
+        }
+
+        if path.is_dir() {
+            std::fs::create_dir_all(&new_path)?;
+            copy_contents(&path, &new_path)?;
+        } else {
+            std::fs::copy(&path, &new_path)?;
```

Comment:
It appears writing contracts and initializing the project is destructive and will overwrite files that already exist. I think we should default to either erroring or skipping files that already exist. The latter would probably be good for us if the intent is to allow rerunning the command with an existing repo to add additional contracts.

Reply:
Great catch! I agree that skipping existing files makes sense - I'll get a change pushed up shortly.

Reply:
updated in [31fdbf9](https://github.com/stellar/soroban-tools/pull/1156/commits/31fdbf999f092fd0d49d14fc812ca4a4635aabe1)